File: day7/cpp/main.cpp
#include <algorithm>
#include <iostream>
#include <map>
#include <memory>
#include <vector>
#include <numeric>
#include <fstream>

using namespace std;

class Directory {
  public:
    shared_ptr<Directory> parent;
    vector<shared_ptr<Directory>> children;
    vector<int> files;
    string name;
    int size;

    Directory(string name_, shared_ptr<Directory> parent_ = nullptr,
              int size_ = 0) {
        this->name = move(name_);
        this->parent = parent_;
        this->size = size_;
    }
};

static map<Directory *, int> MEMO = {};

int getDirectorySize(Directory &directory) {
    if (MEMO.find(&directory) != MEMO.end()) {
        return MEMO.at(&directory);
    }
    int size = 0;
    for (auto file : directory.files) {
        size += file;
    }
    for (auto &child : directory.children) {
        size += getDirectorySize(*child);
    }
    MEMO.insert({&directory, size});
    return size;
}

void first_task(const vector<string> &commands) {
    auto curr_dir = make_shared<Directory>("/");

    for (auto &line : commands) {
        if (line == "$ ls") {
            continue;
        }
        if (line.find("$ cd") != string::npos) {
            string_view target =
                string_view(line.c_str(), line.size()).substr(5);
            if (target == "..") {
                curr_dir = curr_dir->parent;
            } else {
                curr_dir = *std::find_if(
                    begin(curr_dir->children), end(curr_dir->children),
                    [target](auto val) { return val->name == target; });
            }
        } else if (line.find("dir") != string::npos) {
            string_view target =
                string_view(line.c_str(), line.size()).substr(4);
            curr_dir->children.emplace_back(
                make_shared<Directory>(string(target), curr_dir));
        } else if (std::isdigit(line[0])) {
            string_view line_view =
                string_view(line.c_str(), line.size()).substr();
            string_view size_view = line_view.substr(0, line_view.find(' '));
            curr_dir->files.emplace_back(std::stoi(size_view.data()));
        }
    }

    while (curr_dir->parent) {
        curr_dir = curr_dir->parent;
    }

    curr_dir->size = getDirectorySize(*curr_dir);
    std::vector<int> MEMO_values;
    for(auto &item: MEMO){
        if(item.second <= 100000)
            MEMO_values.emplace_back(item.second);
    }
    cout << "First task: " << std::accumulate(begin(MEMO_values), end(MEMO_values), 0) << '\n';
}

void second_task(const vector<string> &commands) {
    auto curr_dir = make_shared<Directory>("/");

    for (auto &line : commands) {
        if (line == "$ ls") {
            continue;
        }
        if (line.find("$ cd") != string::npos) {
            string_view target =
                string_view(line.c_str(), line.size()).substr(5);
            if (target == "..") {
                curr_dir = curr_dir->parent;
            } else {
                curr_dir = *std::find_if(
                    begin(curr_dir->children), end(curr_dir->children),
                    [target](auto val) { return val->name == target; });
            }
        } else if (line.find("dir") != string::npos) {
            string_view target =
                string_view(line.c_str(), line.size()).substr(4);
            curr_dir->children.emplace_back(
                make_shared<Directory>(string(target), curr_dir));
        } else if (std::isdigit(line[0])) {
            string_view line_view =
                string_view(line.c_str(), line.size()).substr();
            string_view size_view = line_view.substr(0, line_view.find(' '));
            curr_dir->files.emplace_back(std::stoi(size_view.data()));
        }
    }

    while (curr_dir->parent) {
        curr_dir = curr_dir->parent;
    }

    curr_dir->size = getDirectorySize(*curr_dir);
    int needed = 30000000 - (70000000 - curr_dir->size);
    int found_min = INT32_MAX;
    auto child = curr_dir;

    for (auto &items : MEMO) {
        if (items.second >= needed) {
            found_min = std::min(found_min, items.second);
        }
    }
    cout << "Second task: " << found_min << '\n';
}

int main(int argc, char *argv[]) {
    ifstream input_file("../input.txt");
    string line;
    vector<string> commands;
    if(!input_file.is_open()){
        cerr << "Couldn't open file\n";
        return 1;
    }
    while(std::getline(input_file, line)){
        commands.emplace_back(move(line));
    }

    first_task(commands);
    second_task(commands);
}

File: day7/rust/src/main.rs
// use std::collections::HashMap;

use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

#[derive(PartialEq, Eq, Debug)]
struct Directory {
    name: String,
    parent: Option<Rc<Directory>>,
    children: RefCell<Vec<Rc<Directory>>>,
    files: RefCell<Vec<i32>>,
}

fn get_directory_size(directory: Rc<Directory>, memo: &mut HashMap<String, i32>) -> i32 {
    let mut key = String::from(&directory.name);
    let mut parent = directory.parent.clone();
    while parent.is_some() {
        let p = parent.clone().unwrap();
        key += &p.name;
        parent = p.parent.clone();
    }

    if memo.contains_key(&key) {
        return *memo.get(&key).unwrap();
    }
    let mut size = 0;
    for file in directory.files.borrow().clone().into_iter() {
        size += file;
    }
    for child in directory.children.borrow().clone().into_iter() {
        size += get_directory_size(child, memo);
    }
    memo.insert(key, size);
    size
}

fn first_task(commands: &[&str]) {
    let root = Rc::new(Directory {
        name: "/".to_string(),
        parent: None,
        children: RefCell::new(vec![]),
        files: RefCell::new(vec![]),
    });
    let mut curr_dir = root.clone();

    for command in commands {
        if *command == "$ ls" {
            continue;
        }
        {
            let target = command.strip_prefix("$ cd").unwrap_or("").trim();
            if !target.is_empty() {
                if target == ".." {
                    curr_dir = curr_dir.clone().parent.clone().unwrap();
                } else {
                    let mut found: Option<Rc<Directory>> = None;
                    {
                        let children = curr_dir.children.borrow_mut();
                        for child in children.clone().into_iter() {
                            if child.name == target {
                                found = Some(child.clone());
                                break;
                            }
                        }
                    }
                    curr_dir = found.unwrap().clone();
                }
                continue;
            }
        }
        {
            let target = command.strip_prefix("dir").unwrap_or("").trim();
            if !target.is_empty() {
                let child = Rc::new({
                    Directory {
                        name: target.to_string(),
                        parent: Some(curr_dir.clone()),
                        children: RefCell::new(vec![]),
                        files: RefCell::new(vec![]),
                    }
                });
                curr_dir.children.borrow_mut().push(child.clone());
                continue;
            }
        }

        if command[0..1].parse::<i32>().is_ok() {
            let size_str = command.split(' ').next().unwrap();
            let size = size_str.parse::<i32>().unwrap();
            curr_dir.files.borrow_mut().push(size);
        }
    }

    let mut memo: HashMap<String, i32> = HashMap::new();
    let _ = get_directory_size(root.clone(), &mut memo);
    let mut result = 0;
    for item in memo.into_iter() {
        if item.1 <= 100000 {
            result += item.1;
        }
    }
    println!("first_task: {}", result);
}

fn second_task(commands: &[&str]) {
    let root = Rc::new(Directory {
        name: "/".to_string(),
        parent: None,
        children: RefCell::new(vec![]),
        files: RefCell::new(vec![]),
    });
    let mut curr_dir = root.clone();

    for command in commands {
        if *command == "$ ls" {
            continue;
        }
        {
            let target = command.strip_prefix("$ cd").unwrap_or("").trim();
            if !target.is_empty() {
                if target == ".." {
                    curr_dir = curr_dir.clone().parent.clone().unwrap();
                } else {
                    let mut found: Option<Rc<Directory>> = None;
                    {
                        let children = curr_dir.children.borrow_mut();
                        for child in children.clone().into_iter() {
                            if child.name == target {
                                found = Some(child.clone());
                                break;
                            }
                        }
                    }
                    curr_dir = found.unwrap().clone();
                }
                continue;
            }
        }
        {
            let target = command.strip_prefix("dir").unwrap_or("").trim();
            if !target.is_empty() {
                let child = Rc::new({
                    Directory {
                        name: target.to_string(),
                        parent: Some(curr_dir.clone()),
                        children: RefCell::new(vec![]),
                        files: RefCell::new(vec![]),
                    }
                });
                curr_dir.children.borrow_mut().push(child.clone());
                continue;
            }
        }

        if command[0..1].parse::<i32>().is_ok() {
            let size_str = command.split(' ').next().unwrap();
            let size = size_str.parse::<i32>().unwrap();
            curr_dir.files.borrow_mut().push(size);
        }
    }

    let mut memo: HashMap<String, i32> = HashMap::new();
    let root_size = get_directory_size(root, &mut memo);
    let needed = 30000000 - (70000000 - root_size);
    let mut found_min = i32::MAX;

    for val in memo.values() {
        if *val >= needed {
            found_min = i32::min(found_min, *val)
        }
    }

    println!("second task: {}", found_min);
}

fn main() {
    let input_file = std::fs::read_to_string("./input.txt").unwrap();
    let commands: Vec<&str> = input_file.lines().collect();
    first_task(&commands);
    second_task(&commands);
}
